warpforge-api/src/formula.rs

```diff
@@ -0,0 +1,41 @@
+use std::ops::Not;
+
+use derive_more::{Display, FromStr};
+use serde::{Deserialize, Serialize};
+
+// FUTURE: Could be represneted as an enum, discriminating on the first char being '/' or '$'
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize, FromStr, Display)]
+pub struct SandboxPort(pub String);
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct GatherDirective {
+	pub from: SandboxPort,
+	pub packtype: Option<crate::content::Packtype>,
+	// TODO:
+	// filters: Option<FilterMap>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum Action {
+	#[serde(rename = "echo")]
+	Echo,
+	#[serde(rename = "exec")]
+	Execute(ActionExecute),
+	#[serde(rename = "script")]
+	Script(ActionScript),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ActionExecute {
+	pub command: Vec<String>,
+	#[serde(default, skip_serializing_if = "<&bool>::not")]
+	pub network: bool,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ActionScript {
+	pub interpreter: String,
+	pub contents: Vec<String>,
+	#[serde(default, skip_serializing_if = "<&bool>::not")]
+	pub network: bool,
```

Comment:
I think this is fine (and probably what golang code is doing), but I wonder if these should be either an `Option<bool>`, or also have a check that forbids deserializing a false.  I'm always a little nervous with things that may implicitly shift serial data during a roundtripping.

Reply:
Changed to `Option<bool>` in https://github.com/janispeyer/rustwarp/commit/33886870d2ff1cc2d17f32fa79334e47d9aa831e